### PR TITLE
Streamline gc alloc interface

### PR DESCRIPTION
**Pull request by densh**
This is a minor change to simplify GC interface to always require an
instance of runtime type information. This is a strict requirement to
support precise garbage collection.

Review @LukasKellenberger

### CODE REVIEW

File: nativelib/src/main/resources/gc/boehm/gc.c [hunk 1]
```diff
@@ -13,12 +14,12 @@ void *scalanative_alloc(void *info, size_t size) {
     return (void *)alloc;
 }
 
-void *scalanative_alloc_raw(size_t size) { return GC_malloc(size); }
-
-void *scalanative_alloc_raw_atomic(size_t size) {
-    return GC_malloc_atomic(size);
+void *scalanative_alloc_atomic(void *info, size_t size) {
+    void **alloc = (void **)GC_malloc_atomic(size);
+    memset(alloc, 0, size);
```
Comment (LukasKellenberger) [inline nativelib/src/main/resources/gc/boehm/gc.c]: Why do we need `memset` here ? Does `GC_malloc_atomic` not guarantee zeroed memory ?

File: nativelib/src/main/resources/gc/boehm/gc.c [hunk 2]
```diff
@@ -13,12 +14,12 @@ void *scalanative_alloc(void *info, size_t size) {
...
+    *alloc = info;
+    return (void *)alloc;
 }
```
Comment (densh) [inline nativelib/src/main/resources/gc/boehm/gc.c]: `alloc_atomic` is still here, left purely not to regress performance with Boehm GC.

File: nativelib/src/main/resources/gc/boehm/gc.c (hunk 1 above)
Comment (densh) [inline nativelib/src/main/resources/gc/boehm/gc.c]: It doesn't! From [the docs](http://www.hboehm.info/gc/gcinterface.html):
```
void * GC_MALLOC_ATOMIC(size_t nbytes)

Allocates nbytes of storage. Requires (amortized) time proportional to nbytes. 
The resulting object will be automatically deallocated when unreferenced. 
The client promises that the resulting object will never contain any pointers. 
The memory is not cleared. ...
```

Comment (LukasKellenberger) [inline nativelib/src/main/resources/gc/boehm/gc.c]: Interesting ! This means that we were using it wrong until now...

Comment (densh) [inline nativelib/src/main/resources/gc/boehm/gc.c]: Previously it was zeroed in https://github.com/scala-native/scala-native/pull/737/files#diff-738f06a13eea511f93ea2f7c590e0320L81. The only difference is that logic has moved to C code now.

Comment (LukasKellenberger) [inline nativelib/src/main/resources/gc/boehm/gc.c]: I see. Makes more sense the way it's done now.

**Pull request merged**